cpp/src/structure/induced_subgraph_impl.cuh

```diff
@@ -102,175 +159,156 @@ extract_induced_subgraphs(
+  //    and dst_subgraph_offsets to use for checking endpoints
+  //    of edges.
+  //
+  rmm::device_uvector<vertex_t> dst_subgraph_vertices_v(0, handle.get_stream());
+  rmm::device_uvector<size_t> dst_subgraph_offsets_v(0, handle.get_stream());
+
+  raft::device_span<size_t const> dst_subgraph_offsets{subgraph_offsets};
+  raft::device_span<vertex_t const> dst_subgraph_vertices{subgraph_vertices};
+
+  auto graph_ids_v =
+    detail::expand_sparse_offsets(subgraph_offsets, vertex_t{0}, handle.get_stream());
 
   if constexpr (multi_gpu) {
-    CUGRAPH_FAIL("Unimplemented.");
-    return std::make_tuple(rmm::device_uvector<vertex_t>(0, handle.get_stream()),
-                           rmm::device_uvector<vertex_t>(0, handle.get_stream()),
-                           rmm::device_uvector<weight_t>(0, handle.get_stream()),
-                           rmm::device_uvector<size_t>(0, handle.get_stream()));
+    auto& comm               = handle.get_comms();
+    auto const comm_rank     = comm.get_rank();
+    auto& row_comm           = handle.get_subcomm(cugraph::partition_2d::key_naming_t().row_name());
+    auto const row_comm_rank = row_comm.get_rank();
+    auto const row_comm_size = row_comm.get_size();
+    auto& col_comm           = handle.get_subcomm(cugraph::partition_2d::key_naming_t().col_name());
+    auto const col_comm_rank = col_comm.get_rank();
+    auto const col_comm_size = col_comm.get_size();
+
+    dst_subgraph_vertices_v = cugraph::device_allgatherv(handle, row_comm, subgraph_vertices);
+
+    graph_ids_v = cugraph::device_allgatherv(
+      handle, row_comm, raft::device_span<vertex_t const>(graph_ids_v.data(), graph_ids_v.size()));
+
+    thrust::sort(handle.get_thrust_policy(),
+                 thrust::make_zip_iterator(graph_ids_v.begin(), dst_subgraph_vertices_v.begin()),
+                 thrust::make_zip_iterator(graph_ids_v.end(), dst_subgraph_vertices_v.end()));
+
+    dst_subgraph_offsets_v =
+      detail::compute_sparse_offsets<size_t>(graph_ids_v.begin(),
+                                             graph_ids_v.end(),
+                                             size_t{0},
+                                             size_t{subgraph_offsets.size() - 1},
+                                             handle.get_stream());
+
+    dst_subgraph_offsets =
+      raft::device_span<size_t const>(dst_subgraph_offsets_v.data(), dst_subgraph_offsets_v.size());
   } else {
-    // 2-1. Phase 1: calculate memory requirements
+    dst_subgraph_vertices_v.resize(graph_ids_v.size(), handle.get_stream());
+    raft::copy(dst_subgraph_vertices_v.data(),
+               subgraph_vertices.data(),
+               subgraph_vertices.size(),
+               handle.get_stream());
+    thrust::sort(handle.get_thrust_policy(),
+                 thrust::make_zip_iterator(graph_ids_v.begin(), dst_subgraph_vertices_v.begin()),
+                 thrust::make_zip_iterator(graph_ids_v.end(), dst_subgraph_vertices_v.end()));
```

Comment:
I started with something like your suggestion.  I could go back to that, if we prefer that approach.

During testing, I discovered that the MG path does not require the input subgraph specifications to be sorted, because we have to sort the data after we shuffle.  By adding this copy/sort we can actually drop the requirement that the input be pre-sorted.  I dropped the check from the expensive checks - although I just observed that I did not update the documentation in `graph_functions.hpp`.

If we think that this extra memory/sorting cost isn't worth it in SG, I can go back to requiring the input to be sorted and just use the provided input in SG.

Reply:
Based on the work below, I think the `graph_ids_v` will still be required for populating the frontier.

Reply:
I see, I will let you make the call. This might be more of a strategic decision than purely technical. This really depends on whether we want to sacrifice little performance/memory footprint of SG path for better usability in both SG & MG paths.

Reply:
I took out the requirement (from the documentation) that the data be sorted and made both paths sort.  We can always revisit this later if we decide the sorting overhead needs to be removed in the SG case.